File: hooks/scripts/on_stop.py
"""Stop hook — POST /capture for the most recent user+assistant turn.

Reads the transcript path from the hook payload, scans the tail of the JSONL
to find the latest user message and the latest assistant message, and
fire-and-forgets `/capture` with a short timeout. Failures are swallowed so
the conversation never errors out on memory issues.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

from _common import add_plugin_scripts_to_path, emit, read_hook_input, session_key

add_plugin_scripts_to_path()

from gateway_client import GatewayClient, breaker_open  # noqa: E402


def _extract_text(message: dict) -> str:
    """Pull plain text out of a Claude transcript message.

    Claude Code transcripts store assistant turns as a list of content blocks
    (`type: text`, `type: tool_use`, ...). We concatenate the text blocks and
    drop the rest.
    """
    content = message.get("content")
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts: list[str] = []
        for block in content:
            if not isinstance(block, dict):
                continue
            if block.get("type") == "text" and isinstance(block.get("text"), str):
                parts.append(block["text"])
        return "\n".join(parts).strip()
    return ""


def _last_turn(transcript_path: str) -> tuple[str, str]:
    """Return (last_user_text, last_assistant_text) from the transcript tail."""
    user_text = ""
    assistant_text = ""
    try:
        path = Path(transcript_path)
        if not path.is_file():
            return "", ""
        # Read in reverse for efficiency on long transcripts.
        for line in reversed(path.read_text("utf-8").splitlines()):
            if not line.strip():
                continue
            try:
                entry = json.loads(line)
            except Exception:
                continue
            msg = entry.get("message") if isinstance(entry.get("message"), dict) else entry
            role = msg.get("role")
            if role == "assistant" and not assistant_text:
                assistant_text = _extract_text(msg)
            elif role == "user" and not user_text:
                user_text = _extract_text(msg)
            if user_text and assistant_text:
                break
    except Exception:
        pass
    return user_text, assistant_text


def main() -> int:
    payload = read_hook_input()
    if breaker_open():
        emit({})
        return 0

    transcript = payload.get("transcript_path") or ""
    user_text, assistant_text = _last_turn(transcript)
    if not user_text or not assistant_text:
        emit({})
        return 0

    sk = session_key(payload)
    try:
        client = GatewayClient(timeout=3)
        client.capture(
            user_content=user_text,
            assistant_content=assistant_text,
            session_key=sk,
            session_id=payload.get("session_id") or "",
        )
    except Exception:
        # Gateway is fire-and-forget on its side; a hook-side failure is fine.
        pass

    emit({})
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception:
        emit({})
        sys.exit(0)

File: hooks/scripts/on_stop.js
#!/usr/bin/env node
/**
 * Stop hook — auto-capture latest turn to local FTS5.
 *
 * After N turns, sets consolidation_due flag for the asyncRewake pipeline.
 */
"use strict";

const fs = require("node:fs");
const nodePath = require("node:path");
const { spawn } = require("node:child_process");
const { addPluginScriptsToPath, readHookInputAsync, emit } = require("./_common.js");
const scriptsDir = addPluginScriptsToPath();
const { extractText } = require(nodePath.join(scriptsDir, "memory_reader.js"));

function lastTurn(transcriptPath) {
  let userText = "";
  let assistantText = "";
  let userUuid = "";
  let assistantUuid = "";
  let gitBranch = "";
  let hadToolUse = false;
  try {
    if (!fs.existsSync(transcriptPath)) return { userText, assistantText, sourceMessageIds: [], gitBranch };
    const lines = fs.readFileSync(transcriptPath, "utf-8").split("\n").reverse();
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      let entry;
      try { entry = JSON.parse(trimmed); } catch { continue; }
      const msg = typeof entry.message === "object" ? entry.message : entry;
      const role = msg.role;
      if (role === "assistant" && !assistantText) {
        assistantText = extractText(msg.content || msg);
        assistantUuid = entry.uuid || "";
        if (Array.isArray(msg.content) && msg.content.some((b) => b && b.type === "tool_use")) hadToolUse = true;
        if (!gitBranch && entry.gitBranch) gitBranch = entry.gitBranch;
      } else if (role === "user" && !userText) {
        userText = extractText(msg.content || msg);
        userUuid = entry.uuid || "";
        if (!gitBranch && entry.gitBranch) gitBranch = entry.gitBranch;
      }
      if (userText && assistantText) break;
    }
  } catch {}
  // source_message_ids is the cross-layer link that was measured DEAD (always []);
  // record the real transcript uuids so consolidate can read the whole turn back.
  const sourceMessageIds = [userUuid, assistantUuid].filter(Boolean);
  return { userText, assistantText, sourceMessageIds, gitBranch, hadToolUse };
}

/**
 * GAP-1 auto-wire: fire-and-forget the deterministic session digest so the durable
 * outcome facts (files edited, tests, releases, git ops) are captured without a
 * skill run. Detached + unref so it never blocks turn-end; idempotent writes (see
 * digest_capture) make running it every tool-turn converge instead of duplicating.
 * Only spawned when the turn actually used a tool — a pure Q&A turn has no new
 * outcomes to capture. Best-effort: any failure is swallowed, memory is optional.
 */
function spawnDigest(transcript, cwd, sessionId) {
  try {
    if (!transcript || !fs.existsSync(transcript)) return;
    const cli = nodePath.join(scriptsDir, "cli.js");
    const child = spawn(process.execPath, [cli, "digest", transcript, "--apply"], {
      cwd: cwd || process.cwd(),
      env: { ...process.env, CLAUDE_PROJECT_DIR: cwd || process.env.CLAUDE_PROJECT_DIR || "" },
      detached: true,
      stdio: "ignore",
    });
    if (child && typeof child.unref === "function") child.unref();
  } catch {}
}

async function main() {
  const payload = await readHookInputAsync();
  const transcript = payload.transcript_path || "";
  const { userText, assistantText, sourceMessageIds, gitBranch, hadToolUse } = lastTurn(transcript);

  // GAP-1: capture machine-certain outcomes from the tool blocks (idempotent).
  if (hadToolUse) spawnDigest(transcript, payload.cwd || "", payload.session_id || "");

  if (userText) {
    try {
      const { autoCapture } = require(nodePath.join(scriptsDir, "memory_auto_capture.js"));
      autoCapture({
        userText,
        assistantText: assistantText || "",
        sessionId: payload.session_id || "",
        cwd: payload.cwd || "",
        sourceMessageIds,
        gitBranch,
        transcriptPath: transcript,
      });
    } catch {}
  }

  emit({});
}

main().catch(() => { emit({}); process.exit(0); });
